### ISSUE TITLE
[docs] Notifications with action buttons on Android

### ISSUE THREAD

**Issue by vron** [expo/expo#10962]
I am not 100% sure if this is a documentation bug or a missing feature.

Trying to schedule local notifications with "quick reply"/action buttons on specific times using https://docs.expo.io/versions/latest/sdk/notifications/

Using

    Notifications.setNotificationCategoryAsync  (specifying action buttons)
    Notifications.scheduleNotificationAsync   (refering the category)

works perfectly on iOS. I cannot get the notifications to show the buttons on Android however (the notification itself shows fine, but without buttons).

Referring to https://docs.expo.io/versions/latest/sdk/notifications/#managing-notification-categories-interactive-notifications there are screenshots from android and the text seems to imply it should be possible to get it to work on Android.

However, referring to https://docs.expo.io/versions/latest/sdk/notifications/#notificationcontent the type specification seems to imply that categoryIdentifier is iOS only.

Is it possible to achieve this in expo on Android? If so, how? I think the documentation is unclear, and I find no obvious way.

(SDK v 39)

**Comment by vron** [comment]
Update,

When running the app as an APK on android the action buttons show up - but not when running it through the expo client.

**Comment by cruzach** [comment]
Please include a reproducible demo if you think there's a bug with `expo-notifications`. Testing this snack- https://snack.expo.io/@charliecruzan/push-notifications - categories are working correctly in the Expo client on both iOS and Android

**Comment by tgxn** [comment]
@cruzach Sorry for reopening this one - I'm trying to test/develop some notification actions on Expo Snack/Android Expo Client and have run into a roadblock with the Snack wokflow:

Correct me if I'm wrong, but the Snack you linked doesn't use notification [`categories`](https://docs.expo.io/versions/latest/sdk/notifications/#managing-notification-categories-interactive-notifications) at all.

This seems to be demonstrated in both iOS and Android but the docs don't mention anywhere that this functionality isn't supported in the expo client.

**Comment by cruzach** [comment]
@tgxn i must have saved over that snack with a new one at some point in the last couple of months- and you're right, I'll fix the docs for that

This screen uses notification categories if you're looking for an example- https://github.com/expo/expo/blob/master/apps/native-component-list/src/screens/NotificationScreen.tsx

**Comment by tgxn** [comment]
@cruzach Cheers for that! It would be nice to have some "official" Snack channel where samples could be stored/updated to demonstrate new features. (as I'm sure, a lot of developers like working from Snacks as a prototype for their application)

It's worth noting that this Issue is still present in Snack with Expo v40.0.0 using the Expo Go client.
_(v41 tested and found Android has [this issue](https://github.com/expo/expo/issues/12833) so unable to validate)_

I've created a Snack to demonstrate this: https://snack.expo.io/@tgxn/b864a5
- Android/Snack environment: Notifications with categories don't function (no button appear)
- Downloaded/Locally: Categories **still** don't work; `android.useNextNotificationsApi` is needed in `app.json` to get categories on Android to function correctly.

**Comment by cruzach** [comment]
It's expected not to work on SDK 40: https://docs.expo.io/push-notifications/sending-notifications/#message-request-format


> categoryId : ID of the notification category that this notification is associated with.&nbsp;Find out more about notification categories here. Must be on at least SDK 41 or bare workflow.

**Comment by leo-r37** [comment]
Hi everyone!. In Expo SDK 50 it still not working.

I configured mi app as next:

`//App.tsx
import { useState, useEffect, useRef } from "react";
import { Text, View, Button, Platform, Alert } from "react-native";
import * as Device from "expo-device";
import * as Notifications from "expo-notifications";
import Constants from "expo-constants";
import { StatusBar } from "expo-status-bar";

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: true,
    shouldSetBadge: false,
  }),
});

// Can use this function below or use Expo's Push Notification Tool from: https://expo.dev/notifications
async function sendPushNotification(expoPushToken) {
  const message = {
    to: expoPushToken,
    sound: "default",
    title: "Mi notificacion",
    body: "Usando expo PUSH notifications!",
    data: { someData: "goes here" },
  };

  await fetch("https://exp.host/--/api/v2/push/send", {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Accept-encoding": "gzip, deflate",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(message),
  });
}

async function registerForPushNotificationsAsync() {
  let token;

  if (Platform.OS === "android") {
    Notifications.setNotificationChannelAsync("default", {
      name: "default",
      importance: Notifications.AndroidImportance.MAX,
      vibrationPattern: [0, 250, 250, 250],
      lightColor: "#FF231F7C",
    });

    Notifications.setNotificationCategoryAsync("emergencia", [
      {
        buttonTitle: "Asistiré",
        identifier: "asiste",
        options: { opensAppToForeground: true },
      },
      {
        buttonTitle: "No asistiré",
        identifier: "no asiste",
        options: { opensAppToForeground: true },
      },
    ]);
  }

  if (Device.isDevice) {
    const { status: existingStatus } =
      await Notifications.getPermissionsAsync();
    let finalStatus = existingStatus;
    if (existingStatus !== "granted") {
      const { status } = await Notifications.requestPermissionsAsync();
      finalStatus = status;
    }
    if (finalStatus !== "granted") {
      alert("Failed to get push token for push notification!");
      return;
    }

    try {
      token = await Notifications.getExpoPushTokenAsync({
        projectId: Constants?.expoConfig?.extra?.eas.projectId,
      });
    } catch (error) {
      console.log("Error al obtener el token", error);
    }

    token && console.log(token);
  } else {
    alert("Must use physical device for Push Notifications");
  }

  return token ? token.data : undefined;
}

export default function App() {
  const [expoPushToken, setExpoPushToken] = useState("");
  const [notification, setNotification] = useState(false);
  const notificationListener = useRef();
  const responseListener = useRef();

  useEffect(() => {
    registerForPushNotificationsAsync().then((token) =>
      setExpoPushToken(token)
    );

    notificationListener.current =
      Notifications.addNotificationReceivedListener((notification) => {
        setNotification(notification);
      });

    responseListener.current =
      Notifications.addNotificationResponseReceivedListener((response) => {
        Alert.alert("Notificacion");

        console.log(response);
      });

    return () => {
      Notifications.removeNotificationSubscription(
        notificationListener.current
      );
      Notifications.removeNotificationSubscription(responseListener.current);
    };
  }, []);

  return (
    <View
      style={{ flex: 1, alignItems: "center", justifyContent: "space-around" }}
    >
      <StatusBar style="dark" />
      <Text>Your expo push token: {expoPushToken}</Text>
      <View style={{ alignItems: "center", justifyContent: "center" }}>
        <Text>
          Title: {notification && notification.request.content.title}{" "}
        </Text>
        <Text>Body: {notification && notification.request.content.body}</Text>
        <Text>
          Data:{" "}
          {notification && JSON.stringify(notification.request.content.data)}
        </Text>
      </View>
      <Button
        title="Press to Send Notification"
        onPress={async () => {
          await sendPushNotification(expoPushToken);
        }}
      />
    </View>
  );
}
`

and a node server with expo-server-sdk:

`import { Request, Response } from "express";
import { Expo } from "expo-server-sdk";

const sendPushNotification = async (req: Request, res: Response) => {
  let expo = new Expo();

  const { tokens, title, body } = req.body;

  if (tokens && tokens.length <= 0) {
    res.status(400).send({ error: "Proveer al menos un token" });
    return;
  }

  let messages = [];
  for (let pushToken of tokens) {
    if (!Expo.isExpoPushToken(pushToken)) {
      console.error(`Push token ${pushToken} is not a valid Expo push token`);
      continue;
    }

    messages.push({
      to: pushToken,
      sound: "default",
      title,
      body,
      categoryId: "emergencia",
    });
  }

  //@ts-ignore
  let chunks = expo.chunkPushNotifications(messages);
  //@ts-ignore
  let tickets = [];

  (async () => {
    for (let chunk of chunks) {
      try {
        let ticketChunk = await expo.sendPushNotificationsAsync(chunk);
        console.log(ticketChunk);
        tickets.push(...ticketChunk);
      } catch (error) {
        console.error(error);
      }
    }
  })();

  let receiptIds = [];
  //@ts-ignore
  for (let ticket of tickets) {
    if (ticket.id) {
      receiptIds.push(ticket.id);
    }
  }

  let receiptIdChunks = expo.chunkPushNotificationReceiptIds(receiptIds);
  (async () => {
    for (let chunk of receiptIdChunks) {
      try {
        let receipts = await expo.getPushNotificationReceiptsAsync(chunk);
        console.log(receipts);

        for (let receiptId in receipts) {
          //@ts-ignore
          let { status, message, details } = receipts[receiptId];
          if (status === "ok") {
            continue;
          } else if (status === "error") {
            console.error(
              `There was an error sending a notification: ${message}`
            );
            //@ts-ignore
            if (details && details.error) {
              // The error codes are listed in the Expo documentation:
              // https://docs.expo.io/push-notifications/sending-notifications/#individual-errors
              // You must handle the errors appropriately.
              //@ts-ignore
              console.error(`The error code is ${details.error}`);
            }
          }
        }
      } catch (error) {
        console.error(error);
      }
    }
  })();

  res.send("Notificacion enviada");
};

export default sendPushNotification;
`

Notifications are delivered correctly but action buttons do not appear.

**Comment by mirsahib** [comment]
Why the issue is closed i am also facing the same issue even after installing the apk into a real device(redmi note 7)

**Comment by dananjaya6005** [comment]
> Hi everyone!. In Expo SDK 50 it still not working.
> 
> I configured mi **app** as next:
> 
> ```js
> //App.tsx
> import { useState, useEffect, useRef } from "react";
> import { Text, View, Button, Platform, Alert } from "react-native";
> import * as Device from "expo-device";
> import * as Notifications from "expo-notifications";
> import Constants from "expo-constants";
> import { StatusBar } from "expo-status-bar";
> 
> Notifications.setNotificationHandler({
>   handleNotification: async () => ({
>     shouldShowAlert: true,
>     shouldPlaySound: true,
>     shouldSetBadge: false,
>   }),
> });
> 
> // Can use this function below or use Expo's Push Notification Tool from: https://expo.dev/notifications
> async function sendPushNotification(expoPushToken) {
>   const message = {
>     to: expoPushToken,
>     sound: "default",
>     title: "Mi notificacion",
>     body: "Usando expo PUSH notifications!",
>     data: { someData: "goes here" },
>   };
> 
>   await fetch("https://exp.host/--/api/v2/push/send", {
>     method: "POST",
>     headers: {
>       Accept: "application/json",
>       "Accept-encoding": "gzip, deflate",
>       "Content-Type": "application/json",
>     },
>     body: JSON.stringify(message),
>   });
> }
> 
> async function registerForPushNotificationsAsync() {
>   let token;
> 
>   if (Platform.OS === "android") {
>     Notifications.setNotificationChannelAsync("default", {
>       name: "default",
>       importance: Notifications.AndroidImportance.MAX,
>       vibrationPattern: [0, 250, 250, 250],
>       lightColor: "#FF231F7C",
>     });
> 
>     Notifications.setNotificationCategoryAsync("emergencia", [
>       {
>         buttonTitle: "Asistiré",
>         identifier: "asiste",
>         options: { opensAppToForeground: true },
>       },
>       {
>         buttonTitle: "No asistiré",
>         identifier: "no asiste",
>         options: { opensAppToForeground: true },
>       },
>     ]);
>   }
> 
>   if (Device.isDevice) {
>     const { status: existingStatus } =
>       await Notifications.getPermissionsAsync();
>     let finalStatus = existingStatus;
>     if (existingStatus !== "granted") {
>       const { status } = await Notifications.requestPermissionsAsync();
>       finalStatus = status;
>     }
>     if (finalStatus !== "granted") {
>       alert("Failed to get push token for push notification!");
>       return;
>     }
> 
>     try {
>       token = await Notifications.getExpoPushTokenAsync({
>         projectId: Constants?.expoConfig?.extra?.eas.projectId,
>       });
>     } catch (error) {
>       console.log("Error al obtener el token", error);
>     }
> 
>     token && console.log(token);
>   } else {
>     alert("Must use physical device for Push Notifications");
>   }
> 
>   return token ? token.data : undefined;
> }
> 
> export default function App() {
>   const [expoPushToken, setExpoPushToken] = useState("");
>   const [notification, setNotification] = useState(false);
>   const notificationListener = useRef();
>   const responseListener = useRef();
> 
>   useEffect(() => {
>     registerForPushNotificationsAsync().then((token) =>
>       setExpoPushToken(token)
>     );
> 
>     notificationListener.current =
>       Notifications.addNotificationReceivedListener((notification) => {
>         setNotification(notification);
>       });
> 
>     responseListener.current =
>       Notifications.addNotificationResponseReceivedListener((response) => {
>         Alert.alert("Notificacion");
> 
>         console.log(response);
>       });
> 
>     return () => {
>       Notifications.removeNotificationSubscription(
>         notificationListener.current
>       );
>       Notifications.removeNotificationSubscription(responseListener.current);
>     };
>   }, []);
> 
>   return (
>     <View
>       style={{ flex: 1, alignItems: "center", justifyContent: "space-around" }}
>     >
>       <StatusBar style="dark" />
>       <Text>Your expo push token: {expoPushToken}</Text>
>       <View style={{ alignItems: "center", justifyContent: "center" }}>
>         <Text>
>           Title: {notification && notification.request.content.title}{" "}
>         </Text>
>         <Text>Body: {notification && notification.request.content.body}</Text>
>         <Text>
>           Data:{" "}
>           {notification && JSON.stringify(notification.request.content.data)}
>         </Text>
>       </View>
>       <Button
>         title="Press to Send Notification"
>         onPress={async () => {
>           await sendPushNotification(expoPushToken);
>         }}
>       />
>     </View>
>   );
> }
> ```
> 
> and a **node server** with expo-server-sdk:
> 
> ```js
> import { Request, Response } from "express";
> import { Expo } from "expo-server-sdk";
> 
> const sendPushNotification = async (req: Request, res: Response) => {
>   let expo = new Expo();
> 
>   const { tokens, title, body } = req.body;
> 
>   if (tokens && tokens.length <= 0) {
>     res.status(400).send({ error: "Proveer al menos un token" });
>     return;
>   }
> 
>   let messages = [];
>   for (let pushToken of tokens) {
>     if (!Expo.isExpoPushToken(pushToken)) {
>       console.error(`Push token ${pushToken} is not a valid Expo push token`);
>       continue;
>     }
> 
>     messages.push({
>       to: pushToken,
>       sound: "default",
>       title,
>       body,
>       categoryId: "emergencia",
>     });
>   }
> 
>   //@ts-ignore
>   let chunks = expo.chunkPushNotifications(messages);
>   //@ts-ignore
>   let tickets = [];
> 
>   (async () => {
>     for (let chunk of chunks) {
>       try {
>         let ticketChunk = await expo.sendPushNotificationsAsync(chunk);
>         console.log(ticketChunk);
>         tickets.push(...ticketChunk);
>       } catch (error) {
>         console.error(error);
>       }
>     }
>   })();
> 
>   let receiptIds = [];
>   //@ts-ignore
>   for (let ticket of tickets) {
>     if (ticket.id) {
>       receiptIds.push(ticket.id);
>     }
>   }
> 
>   let receiptIdChunks = expo.chunkPushNotificationReceiptIds(receiptIds);
>   (async () => {
>     for (let chunk of receiptIdChunks) {
>       try {
>         let receipts = await expo.getPushNotificationReceiptsAsync(chunk);
>         console.log(receipts);
> 
>         for (let receiptId in receipts) {
>           //@ts-ignore
>           let { status, message, details } = receipts[receiptId];
>           if (status === "ok") {
>             continue;
>           } else if (status === "error") {
>             console.error(
>               `There was an error sending a notification: ${message}`
>             );
>             //@ts-ignore
>             if (details && details.error) {
>               // The error codes are listed in the Expo documentation:
>               // https://docs.expo.io/push-notifications/sending-notifications/#individual-errors
>               // You must handle the errors appropriately.
>               //@ts-ignore
>               console.error(`The error code is ${details.error}`);
>             }
>           }
>         }
>       } catch (error) {
>         console.error(error);
>       }
>     }
>   })();
> 
>   res.send("Notificacion enviada");
> };
> 
> export default sendPushNotification;
> ```
> 
> Notifications are delivered correctly but action buttons do not appear.
> 
> I'm using a physical device to test it

DId you found any solution ?

**Comment by kiikoh** [comment]
Anyone have a solution?

**Comment by nikwebr** [comment]
For me, the action buttons do also not show up on an APK build on a physical android device although they do show up on iOS.

I use the following Code in the main entry file and sent push notifications with the expo-server-sdk including the categroyIdentifier `ActionIdentifiers.CHECKIN`:
```
useEffect(() => {
    if (Platform.OS === 'android') {
        await Notifications.setNotificationChannelAsync('default', {
            name: 'default',
            importance: Notifications.AndroidImportance.MAX,
            vibrationPattern: [0, 250, 250, 250],
            lightColor: '#FF231F7C',
        });
    }
    await Notifications.setNotificationCategoryAsync("reminder", [{buttonTitle: "Es ist alles okay", identifier: ActionIdentifiers.CHECKIN, options: {
            opensAppToForeground: false
    }}])
}, []);
```

**Comment by ddabrowiecki** [comment]
Hi all,

After hours of searching and research, I have gotten this to work on my Expo app's dev build. Hope this can help some folks:

The `categoryIdentifier` attribute in the `content` object which you pass to `Notifications.scheduleNotificationAsync()` is the key.

`categoryIdentifier` must match the first parameter passed to `Notifications.setNotificationCategoryAsync()`

So, setting up your notification, first set up your `NotificationChannel` as @leo-r37 mentioned above.  Then, set your Notification category:

```
const actions: Notifications.NotificationAction[] = [
      {
        identifier: "first",
        buttonTitle: "Send a message",
        textInput: {
          placeholder: "Message to send",
          submitButtonTitle: "Send",
        },
      },
      {
        identifier: "second",
        buttonTitle: "Dismiss"
      },
    ];
    // Add your identifier here as the first parameter
    await Notifications.setNotificationCategoryAsync("yourIdentifierHere", actions)
```

Then, when you are scheduling your notification:

```
    Notifications.scheduleNotificationAsync(
      {
        content: {
          title: "test",
          body: "Remember to test!",
          // Make this match the identifier above
          categoryIdentifier: "yourIdentifierHere",
        },
        trigger: { date: dateTime },
      }
    )
```

This worked for me on a dev build and my app only sends **local notifications**.

I found this solution on [this Medium post](https://medium.com/@mvspavs/expo-notifications-171ba5ab951d). The answer is somewhat mentioned in the docs [here](https://docs.expo.dev/versions/latest/sdk/notifications/#manage-notification-categories-for-interactive-notifications) , but I found this section super confusing. I originally thought that setting notification categories was only possible on iOS based on the wording!

Could contributors (pinging @cruzach) consider updating the `README` with some example code in the aforementioned section of the docs?
